### PR TITLE
fix: render forwarded channel search inner messages

### PR DESCRIPTION
**Pull request by LLwill**
## Summary
- map backend channel search forward hits from inner_messages into the frontend model
- render forwarded search results with inner message sender/text rows and fallback placeholders
- update ChannelSearch Storybook mock, i18n strings, and adapter coverage

## Test
- pnpm --filter @octo/base exec vitest run src/Components/ChannelSearch/__tests__
- pnpm i18n:check
- pnpm --filter @octo/web build

### CODE REVIEW

File: packages/dmworkbase/src/Components/ChannelSearch/index.tsx [hunk 1]
```diff
@@ -795,6 +834,17 @@ const MessageResultItem = React.memo(function MessageResultItem({
   const { format, t } = useI18n();
   const sender = resolveSender(item, getSender);
   const isForward = item.kind === "merge_forward";
+  const forwardInnerMessages = item.forward?.innerMessages || [];
+  const visibleForwardInnerMessages = forwardInnerMessages.slice(
+    0,
+    FORWARD_INNER_MESSAGE_DISPLAY_LIMIT
+  );
```
Comment (OctoBoooot) [inline packages/dmworkbase/src/Components/ChannelSearch/index.tsx:807]: **[minor]** `FORWARD_INNER_MESSAGE_DISPLAY_LIMIT = 4` silently truncates `innerMessages` with no `+N more` indicator. Old `snippets.map` rendered all entries, and `childCount` only fires when BOTH `innerMessages` AND `snippets` are empty (`index.tsx:893-898` chain), so a forward with `childCount: 10` and `innerMessages: [10 entries]` shows 4 messages and no count — the user can't tell more were hidden. Either surface `childCount` alongside the visible messages, or render a `+{childCount - visible.length} more` row when the slice trims.

**Review by OctoBoooot — COMMENTED** [review]
## Review: fix: render forwarded channel search inner messages (#437)

**Verdict:** Approve with comments

Scope is clean (7 files, +188/-12), CI green, risk tier normal. Adapter mapping has BMP + missing-field coverage in `apiAdapter.test.ts:351-400`, types are tightened end-to-end, and the i18n keys are mirrored across en-US/zh-CN. One minor UX note inline; otherwise this is a focused fix.

### Praise
- `resolveForwardInnerMessageSenderName` (`index.tsx:94-101`) gets a subtle case right: `channelSearchEmptyDataSource.getSender(uid)` returns `{uid, name: uid}` for unknown senders (`adapter.ts:5-8`), so the `sender.name !== message.senderUid` guard correctly suppresses the `uid：text` UI that would otherwise leak the raw id. Combined with `formatForwardInnerMessage`'s `senderName ? ${name}：${text} : text` (`index.tsx:122`), an unknown sender renders as just the message body instead of `c3060f44...：[Image]`. The placeholder fallback chain (`text → forwardInnerMessageFallbackText(type)`) using `MessageContentTypeConst` (`index.tsx:104-117`) is similarly defensive — an empty `search_text` for a non-text type still renders as a typed `[Image]`/`[Video]`/`[File]` rather than a blank row.

**Review by mochashanyao — DISMISSED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-web PR#437 Review Report
**Reviewer**: Octo-Q (automated review)
**PR:** Mininglamp-OSS/octo-web#437
**Head SHA:** `fa7b5e840784d54e0866a26cb57b68118b40ef27`
**Title:** fix: render forwarded channel search inner messages

---

## 1. Verification Summary

| Item | Status | Evidence |
|------|--------|----------|
| Backend → adapter mapping | ✅ | `apiAdapter.ts:258-270` — `mapForwardInnerMessage` correctly maps all 6 fields with safe defaults |
| `innerMessages` optional handling | ✅ | `apiAdapter.ts:305-307` — `Array.isArray` guard; `index.tsx:838` — `\|\| []` fallback |
| Fallback text for empty `search_text` | ✅ | `index.tsx:107-118` — correctly checks `MessageContentTypeConst.image(2)/smallVideo(5)/file(8)` from `Const.ts:50-58` |
| Sender name resolution | ✅ | `index.tsx:93-100` — 3-tier fallback: `senderName` → `getSender(uid).name` → `""` |
| `<mark>` tag handling in HighlightText | ✅ | `index.tsx:233-268` — React text-node rendering prevents XSS; pre-existing behavior unchanged |
| Fallback cascade (innerMessages → snippets → childCount) | ✅ | `index.tsx:886-910` — mutually exclusive conditions, correct priority |
| i18n key parity (en-US / zh-CN) | ✅ | Both locales add 5 matching keys with consistent `{{keyword}}` / `{{count}}` interpolation |
| Test coverage | ✅ | `apiAdapter.test.ts:342-399` — covers mapping with inner messages, empty text, missing sender_name, missing sent_at |
| Storybook mock update | ✅ | `ChannelSearch.stories.mock.ts:155-183` — migrated from `snippets` to `innerMessages`; `itemMatchesKeyword` updated |
| Type definitions | ✅ | `types.ts:56-63` — `ChannelSearchForwardInnerMessage` interface matches adapter output |

## 2. Findings

### No P0 or P1 issues found.

### P2 Observations

**P2-1: React key uses index fallback**
- **Location:** `index.tsx:889` — `key={\`${message.messageId || "inner"}-${index}\`}`
- **Diff-scope:** new (introduced by this PR)
- **Severity:** P2 / nit
- When `messageId` is empty string (from `hit.message_id || ""`), all messages without IDs share the `"inner"` prefix and rely on index. Acceptable for this static render (no reorder/insert/delete), but would cause reconciliation issues if the list ever becomes dynamic.

**P2-2: `forwardMatchReason` always overrides backend matchReason for forward hits**
- **Location:** `index.tsx:843-848`
- **Diff-scope:** new (introduced by this PR)
- **Severity:** P2 / product decision
- When `isForward && keyword.trim()` is truthy, the backend-provided `item.matchReason` is replaced with a synthesized `"Forwarded chat contains \"keyword\""`. This is a reasonable UX choice (the real match is in inner messages, not the outer snippet), but note that any backend-specific match reason for forward hits is discarded.

## 3. Suggestions

- Consider extracting `MessageContentTypeConst` values used in `forwardInnerMessageFallbackText` into a local set/lookup if more message types are added in the future (currently 3 if-checks, manageable).
- The `optionalSentAtToSeconds` helper is clean and well-tested. Consider reusing it in other adapter paths that currently use `sentAtToSeconds` with its `Date.now()` fallback, where `undefined` would be more appropriate.

## 4. Additional Findings

- No security concerns. `HighlightText` renders all content through React text nodes — `<mark>` tags are parsed by regex and content is never injected as raw HTML. This applies equally to the new inner message text path.
- The `I18nContextValue` type import is correctly resolved via `i18n/index.ts` re-export.

## 5. Data Flow Trace

```
Backend API response
  └─ MessageSearchHit.inner_messages?: ForwardInnerMessageHit[]
       ├─ message_id → messageId (|| "")
       ├─ type → type (number check, default 0)
       ├─ search_text → text (|| "")
       ├─ sender_id → senderUid (|| undefined)
       ├─ sender_name → senderName (|| undefined)
       └─ sent_at → timestamp (optionalSentAtToSeconds → undefined on missing/invalid)
  ↓ (only when message_kind === "forward")
ChannelSearchItem.forward.innerMessages?: ChannelSearchForwardInnerMessage[]
  ↓
MessageResultItem rendering:
  ├─ forwardInnerMessages = item.forward?.innerMessages || []
  ├─ visibleForwardInnerMessages = .slice(0, 4)
  ├─ For each visible message:
  │    ├─ text = message.text || forwardInnerMessageFallbackText(type, t)
  │    │    └─ type 2→[Image], 5→[Video], 8→[File], else→[Message]
  │    ├─ senderName = message.senderName || getSender(uid).name || ""
  │    └─ formatted = senderName ? "senderName：text" : text
  ├─ If no innerMessages → fall back to snippets
  └─ If no snippets → fall back to childCount display
```

All data paths verified: no empty-array short-circuits, no premature returns, no gate-controlled bypasses that would prevent inner messages from reaching the render point.

## 6. Blind Spot Checklist (R5)

- **C1 — Dual-path parity:** N/A. No symmetric add/remove or create/delete paths touched.
- **C2 — Control-flow ordering / nested reuse:** Clear. `HighlightText` is reused but the new `formatForwardInnerMessage` output is a plain string consumed identically to existing snippet text. No double-processing or ordering issues.
- **C3 — Authorization boundary:** N/A. No auth/permission changes.
- **C4 — Authorization lifecycle:** N/A. No auth changes.
- **C5 — Build ≠ runtime:** N/A. No build/packaging/extension changes.
- **C6 — Governance docs:** N/A. No documentation changes.

## 7. Cross-round Blocker Recheck (R6)

N/A — first review of this PR.

---

[Octo-Q] verdict: **APPROVE**

No P0/P1 issues. The PR correctly maps backend `inner_messages` into the frontend model, handles all edge cases (empty text, missing sender, missing timestamp, absent inner_messages array), and provides proper fallback cascading. Test coverage is adequate. The two P2 observations are minor and non-blocking.

**Review by Jerry-Xin — CHANGES_REQUESTED** [review]
In-scope PR, but the forwarded inner-message rendering can duplicate the sender name for the backend shape added by this PR.

**🔴 Blocking**
- 🔴 Critical: Forwarded inner messages can render with a duplicated sender label. The adapter preserves `inner_messages[].search_text` verbatim as `text` in [apiAdapter.ts](/Users/mlamp/.openclaw/tmp/pr-review-wvf0ia0l/repo/packages/dmworkbase/src/Components/ChannelSearch/apiAdapter.ts:267), and the updated adapter test documents a realistic value where `search_text` already includes the sender label in [apiAdapter.test.ts](/Users/mlamp/.openclaw/tmp/pr-review-wvf0ia0l/repo/packages/dmworkbase/src/Components/ChannelSearch/__tests__/apiAdapter.test.ts:359). The UI then unconditionally prepends `senderName` in [index.tsx](/Users/mlamp/.openclaw/tmp/pr-review-wvf0ia0l/repo/packages/dmworkbase/src/Components/ChannelSearch/index.tsx:127) before rendering at [index.tsx](/Users/mlamp/.openclaw/tmp/pr-review-wvf0ia0l/repo/packages/dmworkbase/src/Components/ChannelSearch/index.tsx:889). That means a hit like sender `Alice` plus text already starting with `Alice:` renders as `Alice: Alice: ...`. Please normalize this contract in one place: either strip the sender prefix in the adapter, or only prepend when the text does not already include it. Add a render-level test or exported formatter test so this does not regress.

**💬 Non-blocking**
- 🔵 Suggestion: The current coverage is adapter-only, so it cannot catch UI formatting regressions like the duplicated sender label. A focused test for forwarded search result rendering would cover sender fallback, placeholder fallback, and highlighted inner text together.

**✅ Highlights**
- The backend `inner_messages` mapping is scoped to forwarded message hits and keeps timestamps optional.
- Locale keys were added for both required baseline locales.

Verification note: I attempted `pnpm --filter @octo/base exec vitest run src/Components/ChannelSearch/__tests__`, but this checkout reports `Command "vitest" not found`, so I could not independently run the focused tests here.

**Review by yujiawei — DISMISSED** [review]
# Code Review — PR #437 (octo-web)

**Scope reviewed:** head `fa7b5e840784d54e0866a26cb57b68118b40ef27` against merge-base `c2ea912b8455d653defc142b2417ea20042b4e68` (target `main`). 7 files, +188/-12.

## 1. Summary

The PR maps backend `inner_messages` from forwarded (merge-forward) channel-search hits into the frontend model and renders each child message as a `sender：text` row, with localized placeholders for non-text payloads and a graceful fallback chain to the old `snippets` / `childCount` display. The change is well-structured, type-safe, and the new behavior is additive (old `snippets` path is preserved as a fallback).

I cross-checked the suspected high-severity issues against the actual backend contract (search-indexer + octo-server). **Both candidate blockers turned out to be false positives** driven by an unrealistic unit-test fixture and an assumption about the i18n API. Details below.

## 2. Verification results

- ✅ **Build passes**: `pnpm --filter @octo/web build` → `✓ built in 4.28s` (only pre-existing, unrelated rolldown warnings).
- ✅ **Tests pass**: `pnpm --filter @octo/base exec vitest run src/Components/ChannelSearch/__tests__` → 37 passed.
- ✅ **i18n check passes**: `pnpm i18n:check` → healthy, both `en-US` and `zh-CN` keys added symmetrically.
- ✅ **No HTML-injection risk**: `HighlightText` renders React text nodes, not `dangerouslySetInnerHTML`; prepending `senderName` cannot inject markup.

## 3. Investigated and cleared (would have been blockers if true)

### Double sender-name prefix — NOT a real bug
`formatForwardInnerMessage` (`index.tsx:127`) prepends `${senderName}：${text}`. The adapter unit-test fixture (`apiAdapter.test.ts:359`) feeds `search_text: "Alice：..."` *together with* `sender_name: "Alice"`, which would render `Alice：Alice：...`. I traced the real contract:

- Indexer `buildSearchText` (`octo-search-indexer/internal/esindex/buildraw.go:244`) joins **only content/caption/name fields** for inner messages — no sender prefix.
- Server `buildInnerMessages` (`octo-server/modules/messages_search/source.go:176`) copies that raw `searchText` into `search_text` and fills `sender_name` **separately** via `senderJoin` after the fact.

So in production `search_text` is body-only and the single frontend prefix is correct. **The test fixture is the thing that's wrong**, not the component. Recommend fixing the fixture so it reflects the real wire shape (body-only `search_text`), otherwise it bakes in a contract the backend never emits and will mislead the next reader. Non-blocking.

### i18n `{ values: { keyword } }` nesting — correct for this codebase
This project does **not** use react-i18next. `I18nService.interpolate` (`i18n/I18nService.ts:43,140`) reads `options.values`, and the `{ values: { ... } }` wrapper is the established pattern across the file (`index.tsx:908, 1358, 1692`). The new `matchReason` call follows it exactly. No literal `{{keyword}}` will leak.

## 4. Findings (all P2 / advisory — none block merge)

- **[P2] Silent truncation with no overflow indicator** — `index.tsx:837-846`. `FORWARD_INNER_MESSAGE_DISPLAY_LIMIT = 4` slices inner messages, but the `childCount` row only renders when *both* `innerMessages` and `snippets` are empty (`index.tsx:893-898`). A forward with 10 children shows 4 rows and no hint that 6 are hidden. The old `snippets.map` rendered all entries, so this is a mild regression in information density. Suggest a `+{childCount - visible.length} more` row (or surfacing `childCount`) when the slice trims. *(Independently flagged by another reviewer on the PR.)*

- **[P2] Forward match-reason discards the backend excerpt** — `index.tsx:841-846`. For forwards with a keyword, `item.matchReason` (which the adapter populates from the backend `<mark>`-highlighted snippet, `apiAdapter.ts:300`) is always replaced by generic copy ("Forwarded chat contains \"X\""). This loses the exact matched fragment. It's a defensible UX trade (the matched content now shows in the inner-message rows with client-side highlighting), but worth a conscious confirmation that the generic reason is intended.

- **[P2] Hardcoded full-width colon couples all locales to CN typography** — `index.tsx:128`. `${senderName}：${text}` uses U+FF1A unconditionally, so an English locale renders `Alice：[Image]`. This mirrors the *pre-existing* snippet convention (the old mock used `牛爷爷：…`), so it's not a regression, but since this PR localizes the placeholders it'd be consistent to also localize the join separator. Low priority.

## 5. Coverage notes (what I could not fully verify)

- **No render-level test** exists for `formatForwardInnerMessage` / the truncation / the fallback chain — only the adapter mapping is unit-tested. The UI rendering path is unverified by the suite.
- **Sender names depend on indexer fields not yet emitted.** Per `octo-server/modules/messages_search/source.go:84-90`, the inner `from` / `timestamp` fields are forward-compat ("the indexer will start writing in a follow-up release"). Until that ships, `sender_id` is omitted → `senderUid` is undefined → inner rows render text-only with no sender prefix. This is graceful degradation, not a defect, but the sender-name part of this feature is dormant until the indexer follow-up lands.
- I did not exercise a live backend payload end-to-end; conclusions on the wire shape are from reading the indexer + server source, not from a captured production response.

## 6. Advisory legs

This review absorbed two independent advisory passes (a line-level pass and a long-context cross-repo pass). Both surfaced one apparent blocker each (the double-prefix and the i18n nesting); I refuted both against the backend source as documented in section 3. Their surviving signal is the P2 set above. Both legs ran successfully.

## Verdict

No P0/P1 blockers. The change builds, tests green, i18n healthy, and the two scary-looking issues are artifacts of an unrealistic test fixture and an i18n-API assumption, not real defects. The remaining items are P2 polish (overflow indicator, match-reason excerpt, colon localization) plus a fixture cleanup — none gate merge.

Approving. Please consider the overflow-indicator P2 and the misleading adapter test fixture as fast follow-ups.

**Comment by yujiawei** [comment]
Adding backend-source evidence on the duplicated-sender-label concern, since it's currently the blocking item.

I traced the wire contract end to end and the production `search_text` for forwarded inner messages is **body-only — it does not carry a sender prefix**:

- Indexer: `buildSearchText` joins only content/caption/name fields for each inner `msgs[]` entry; no sender label is concatenated (`octo-search-indexer/internal/esindex/buildraw.go`, `buildSearchText` / `searchableTextFields`).
- Server: `buildInnerMessages` copies that raw `searchText` straight into `search_text`, and `sender_name` is populated **separately** via `senderJoin` afterwards (`octo-server/modules/messages_search/source.go`, see the `sender_name is left empty here` note + `buildMessageHits`).

So in production, prepending `senderName` once produces a single, correct prefix. The `Alice：Alice：…` case only arises from the unit-test fixture at `apiAdapter.test.ts:359`, which feeds `search_text: "Alice：…"` together with `sender_name: "Alice"` — a shape the backend never emits. My read is that the **fixture is wrong**, not the component.

That said, the underlying suggestion still has merit as hardening: the unconditional prepend is fragile if the backend contract ever changes, and the suggested render-level/formatter test is worth adding. I'd frame those as P2 follow-ups (fix the fixture to match the real wire shape + add a render test) rather than a correctness blocker, since today's backend can't trigger the duplication.

Happy to be corrected if there's a captured production payload that shows `search_text` arriving pre-prefixed — that would flip this.

**Review by OctoBoooot — APPROVED** [review]
## Re-review: fix: render forwarded channel search inner messages (#437)

**Verdict:** Approve (delta on `96f73044`)

Both peer-flagged issues addressed in code with adversarially-verifiable test coverage. CI Build still in progress, other checks green, risk tier normal.

### Delta against `fa7b5e8`
- **🔴 sender-prefix duplication (Jerry-Xin / Steve / yujiawei — and I missed in round 1)** — new `hasForwardInnerMessageSenderPrefix` (`forwardInnerMessage.ts:46-53`) detects when `text` already starts with `<senderName>\s*[:：]` and short-circuits the prepend so `formatForwardInnerMessage` no longer produces `Alice：Alice：…`. Escapes regex metachars in the sender name (`escapeRegExp` at `forwardInnerMessage.ts:18-20`) so a name like `Al.ice` or `Bob|Alice` matches literally. Handles both half-width `:` and full-width `：`, plus whitespace tolerance via `\s*`. Test cases at `forwardInnerMessage.test.ts:36-62` cover both colon widths. The adapter test fixture was also corrected (`apiAdapter.test.ts:359, 386`) from `Alice：命中的…` to `命中的…` — documenting that the backend `search_text` contract is body-only, while the rendering layer stays resilient if a deployment ever sends a prefixed value.
- **My minor on silent truncation** — `getForwardInnerMessageHiddenCount` (`forwardInnerMessage.ts:69-78`) handles both "more in the array than visible" and "backend `childCount > array length`" via `Math.max(0, total - visible)`; new "+N more chat messages" row renders when `hiddenCount > 0` (`index.tsx:865-871`), with `channelSearch.forward.more` keys mirrored in en-US/zh-CN.
- **Refactor** — helpers extracted from `index.tsx` into `forwardInnerMessage.ts` with dedicated unit tests; the component now imports rather than co-defines, cutting the formatting surface from the React file.

### Praise
- The sender-prefix fix uses defense-in-depth correctly: (a) corrects the test fixture to pin the backend contract as body-only, AND (b) keeps `hasForwardInnerMessageSenderPrefix` as the rendering-layer guard. Either alone would leave a gap — fixing only the fixture trusts the backend forever; fixing only the renderer hides the contract drift. Doing both makes the contract explicit and the UI resilient.
- The `hasForwardInnerMessageSenderPrefix` regex (`forwardInnerMessage.ts:46-53`) is correctly built — `escapeRegExp` on the sender name (so `Al.ice`/`Bob|Alice` don't blow up), `\s*` for whitespace tolerance, `[:：]` for both half- and full-width colons. The test cases at `forwardInnerMessage.test.ts:36-62` cover the two colon widths concretely rather than asserting just one and hoping the other works.

**Review by mochashanyao — APPROVED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-web PR#437 Review Report — `fix: render forwarded channel search inner messages`
**Reviewer**: Octo-Q (automated review)
**Head SHA:** `96f73044f03c03bb4979a812fe8fa486bf209464`
**Repo:** Mininglamp-OSS/octo-web
**CI:** All checks green (10 pass, 5 skipped)

---

## 1. Verification Summary

| Item | Status | Evidence |
|------|--------|----------|
| Type model extension | ✅ | `types.ts` adds `ChannelSearchForwardInnerMessage` interface + optional `innerMessages?` on `ChannelSearchForwardInfo` — backward-compatible |
| Adapter mapping | ✅ | `apiAdapter.ts:mapForwardInnerMessage` correctly maps `inner_messages[]` from backend hit; `mapMessageHit` populates `innerMessages` only when `messageKind === "forward"` |
| Rendering logic | ✅ | `index.tsx:MessageResultItem` renders inner messages with 4-item display limit, "more" overflow, and snippet/childCount fallback |
| Sender name dedup | ✅ | `forwardInnerMessage.ts:hasForwardInnerMessageSenderPrefix` uses escaped regex + both colon variants (`:` / `：`) |
| Fallback text | ✅ | `forwardInnerMessageFallbackText` maps type constants (image=2, smallVideo=5, file=8) to i18n placeholders; default `[Message]` |
| i18n coverage | ✅ | Both `en-US.json` and `zh-CN.json` add 6 new keys (matchReason, more, 4 placeholders) |
| Test coverage | ✅ | New `forwardInnerMessage.test.ts` (98 lines, 4 test cases); updated `apiAdapter.test.ts` with inner_messages mapping verification |
| Storybook mock | ✅ | `ChannelSearch.stories.mock.ts` updated from `snippets` to `innerMessages`; `itemMatchesKeyword` extended to search inner message text/sender |

## 2. Findings

### No P0/P1 issues found.

### P2 — Observations (non-blocking)

**P2-1: `forwardMatchReason` replaces backend `matchReason` for all forward hits with non-empty keyword**
- **Diff-scope:** new (introduced by this PR)
- `index.tsx:815-819`: When `isForward && keyword.trim()`, the component generates an i18n string `"Forwarded chat contains \"keyword\""` instead of using `item.matchReason` from the backend.
- This is an intentional UX improvement (localized, consistent forward-card description). The old `matchReason` for forwards was a raw text snippet that might leak inner-message content.
- **Impact:** Low — behavioral change is deliberate and improves i18n support. No action needed.

**P2-2: `hiddenForwardInnerMessageCount` uses `max(childCount, innerMessages.length)` as total**
- **Diff-scope:** new
- `forwardInnerMessage.ts:72-77`: If `childCount > innerMessages.length`, uses `childCount` as total. This is correct when `childCount` = total messages in the forwarded conversation and `innerMessages` = search-hit subset.
- Edge case: if backend ever returns `childCount < innerMessages.length` (data inconsistency), the function gracefully falls back to `innerMessages.length`. `Math.max(0, ...)` prevents negative counts.
- **Impact:** Low — defensive design handles edge cases well.

**P2-3: React key uses `messageId || "inner"` with index fallback**
- **Diff-scope:** new
- `index.tsx:858`: `key={${message.messageId || "inner"}-${index}}`. If multiple messages lack `messageId`, they all get unique keys via the index suffix. This is fine for a static list (no reordering/insertion), which is the case here.
- **Impact:** Negligible.

## 3. Suggestions

None blocking. The implementation is clean and well-structured.

## 4. Extra Findings

None beyond the diff scope.

## 5. Data-Flow Trace

| Consumed Data | Upstream Source | Flows Correctly? |
|---|---|---|
| `item.forward.innerMessages` | Backend `inner_messages[]` → `apiAdapter.ts:mapMessageHit` → `mapForwardInnerMessage` per hit → `ChannelSearchForwardInnerMessage[]` | ✅ Array.isArray guard; undefined when absent |
| `message.text` (inner) | Backend `search_text` (may contain `<mark>` tags) → `hit.search_text \|\| ""` | ✅ `HighlightText` parses `<mark>` tags correctly after sender-name prepend |
| `message.senderName` (inner) | Backend `sender_name` → `hit.sender_name \|\| undefined`; fallback via `getSender(uid)` → sender cache or `{uid, name:uid}` | ✅ `resolveForwardInnerMessageSenderName` tries inline name first, then cache lookup, then empty string |
| `message.type` (inner) | Backend `type` number → `typeof hit.type === "number" ? hit.type : 0` | ✅ Default 0 maps to generic `[Message]` fallback in `forwardInnerMessageFallbackText` |
| `message.timestamp` (inner) | Backend `sent_at` ISO string → `optionalSentAtToSeconds` → epoch seconds or undefined | ✅ NaN guard returns undefined |
| `forwardMatchReason` | Generated from `keyword.trim()` via i18n `t()` | ✅ Falls back to `item.matchReason` when not forward or keyword empty |
| `visibleForwardInnerMessages` | `.slice(0, FORWARD_INNER_MESSAGE_DISPLAY_LIMIT)` where limit=4 | ✅ Deterministic, no off-by-one |
| `hiddenForwardInnerMessageCount` | `getForwardInnerMessageHiddenCount(innerMessages.length, visibleCount, childCount)` | ✅ `Math.max(0, total - visible)` prevents negative |
| Snippet fallback path | Gated on `visibleForwardInnerMessages.length === 0` | ✅ Only renders old snippets when no inner messages exist (backward compat) |

## 6. Blind-Spot Checklist (R5)

- **C1 — Dual-path parity:** N/A. No symmetric add/remove or create/delete paths touched.
- **C2 — Control-flow ordering / nesting:** N/A. `formatForwardInnerMessage` is called once per inner message in a flat `.map()`. No nesting or double-application risk.
- **C3 — Authorization boundary:** N/A. Pure UI rendering change; no new endpoints or capability exposure.
- **C4 — Authorization lifecycle:** N/A. No auth changes.
- **C5 — Build ≠ runtime:** N/A. Standard React component code; no relative URLs, extension packaging, or environment-dependent paths.
- **C6 — Governance/policy docs:** N/A. No documentation or policy changes.

## 7. Cross-Round Blocker Recheck (R6)

N/A — first review of this PR.

---

**[Octo-Q] verdict: APPROVE**

Clean, well-tested feature addition. The data flow from backend `inner_messages` through adapter mapping to UI rendering is correct and complete. Backward compatibility is preserved via the snippet/childCount fallback path. No P0/P1 issues found.

**Review by Jerry-Xin — APPROVED** [review]
The PR is in scope for `Mininglamp-OSS/octo-web`: it extends the existing ChannelSearch feature to support forwarded inner-message search results.

💬 Non-blocking

🔵 Suggestion: `packages/dmworkbase/src/Components/ChannelSearch/index.tsx:813` now replaces the backend-provided forward `matchReason` with a generic keyword-based string whenever the keyword is non-empty. That is reasonable for the new UX, but it means backend-highlighted wording is no longer surfaced in the match-reason row. If backend snippets carry useful normalized/highlighted context, consider preferring `item.matchReason` when present.

✅ Highlights

The adapter mapping in `apiAdapter.ts:261` and `apiAdapter.ts:301` cleanly keeps inner forwarded messages in the frontend model without disrupting existing snippet fallback behavior.

The render path in `index.tsx:854` safely routes forwarded rows through `HighlightText`, which tokenizes `<mark>` tags into React nodes instead of using raw HTML.

The new helper coverage in `forwardInnerMessage.test.ts:21` covers sender prefixing, placeholder fallback, raw-id suppression, and hidden-count behavior.

Test note: I attempted to rerun `pnpm --filter @octo/base exec vitest run src/Components/ChannelSearch/__tests__`, but this worktree does not have `vitest` installed, so the command could not start.

**Review by yujiawei — APPROVED** [review]
# Code Review — PR #437 (octo-web)

**Scope reviewed:** head `96f73044f03c03bb4979a812fe8fa486bf209464` against merge-base with `main`. 9 files, +343/-11. This is a re-review of the hardening commit `96f7304` that builds on the original `fa7b5e8`.

## 1. Summary

The PR maps backend `inner_messages` from forwarded (merge-forward) channel-search hits into the frontend model and renders each child message as a `sender：text` row, with localized placeholders for non-text payloads, a "+N more" overflow indicator, and a graceful fallback chain to the older `snippets` / `childCount` display. The latest commit extracts the formatting logic into a dedicated, unit-tested `forwardInnerMessage.ts` module and adds sender-prefix de-duplication.

The change is well-structured, type-safe, and additive — the legacy `snippets` path is preserved as a fallback.

## 2. Verification results (run locally on the head SHA)

- ✅ **Tests pass**: `pnpm --filter @octo/base exec vitest run src/Components/ChannelSearch/__tests__` → 41 passed (up from 37; +4 new `forwardInnerMessage` formatter tests).
- ✅ **i18n check passes**: `pnpm i18n:check` → healthy; `en-US` and `zh-CN` keys added symmetrically.
- ✅ **Build passes**: `pnpm --filter @octo/web build` → built clean (only pre-existing, unrelated rolldown warnings).
- ✅ **No HTML-injection risk**: `HighlightText` renders React text nodes / parses `<mark>` into elements, never `dangerouslySetInnerHTML`; prepending `senderName` cannot inject markup.

## 3. Resolved from prior rounds

- ✅ **Duplicated sender-label prefix** — the new `formatForwardInnerMessage` only prepends `senderName` when the text does not already start with it (`hasForwardInnerMessageSenderPrefix`, `forwardInnerMessage.ts:44`), and the misleading adapter test fixture was corrected to the real body-only wire shape (`apiAdapter.test.ts`). A dedicated formatter test now guards this.
- ✅ **No overflow indicator** — a `+N more` row was added (`index.tsx:865`) with `getForwardInnerMessageHiddenCount`.

## 4. Findings

All findings below are **P2 / advisory** — none block merge.

- **[P2] Duplicate count lines when the backend sends `child_count` but no `inner_messages`.** `index.tsx:865` (`+N more`) and `index.tsx:881-889` (`childCount` fallback) are not mutually exclusive. When `innerMessages` is empty/undefined while `childCount > 0`, `getForwardInnerMessageHiddenCount(0, 0, childCount)` returns `childCount`, so the "N more chat messages" row renders, and the existing total-count fallback ("N chat messages") *also* renders — two near-identical count lines. This is the forward-compat / old-backend degradation window (where inner messages are not yet emitted). Suggested fix: gate the `+N more` row on `visibleForwardInnerMessages.length > 0`, so the child-count-only path keeps the single existing line.

- **[P2] Forward match-reason discards the backend excerpt.** `index.tsx:813-818`. For forwards with a keyword, `item.matchReason` (populated from the backend `<mark>`-highlighted snippet at `apiAdapter.ts:300`) is always replaced by generic copy ("Forwarded chat contains \"X\""). If a match is on the forward title / a nested filename rather than inner chat text, the exact matched fragment is no longer surfaced in the reason line. This is a defensible trade now that the matched content shows in the inner-message rows with client-side highlighting, but worth a conscious confirmation that the generic reason is intended.

- **[P2] Prefix-dedup is whitespace/colon-anchored only.** `hasForwardInnerMessageSenderPrefix` (`forwardInnerMessage.ts:44`) anchors on `^senderName\s*[:：]`. The current backend contract emits body-only `search_text` (sender populated separately via `senderJoin`), so a pre-prefixed or `<mark>`-wrapped sender label cannot occur today and this is not a live defect. Keep it in mind if the wire contract ever starts embedding the sender into `search_text`.

- **[P2] `optionalSentAtToSeconds` relies on `Date` parsing of `sent_at`.** `apiAdapter.ts:133`. Non-ISO-8601 inputs (e.g. space-separated SQL-style timestamps) parse to `NaN` in some engines, dropping the value to `undefined`. The inner-message timestamp is not currently rendered, so impact is nil today; this mirrors the pre-existing `sentAtToSeconds` pattern.

## 5. Highlights

- Formatting logic extracted to a focused, unit-tested module — good separation and the new tests cover sender fallback, placeholder fallback, prefix-dedup, and hidden-count math.
- Inner-message mapping is scoped to forward hits, timestamps are optional, and locale keys were added symmetrically for both baseline locales.

## 6. Coverage notes

- This review absorbed two independent advisory passes (a line-level pass and a long-context cross-repo pass). The line-level pass surfaced the duplicate-count-line P2 (independently verified by a logic reproduction). The long-context pass flagged the match-reason override and the prefix-dedup edge as higher severity; I down-weighted both to P2 after tracing the backend contract (body-only `search_text`) and confirming the intentional UX trade. Two of its candidates (HighlightText tag corruption; snippet suppression) were refuted — `HighlightText` treats `<mark>`-bearing text and keyword-matching as mutually exclusive, and the adapter always sets `snippets: []` for forward hits, so neither can occur in the current code. Both advisory legs ran successfully.
- No live backend payload was exercised end-to-end; wire-shape conclusions are from reading the indexer + server source plus the adapter tests.

## Verdict

No P0/P1 blockers. The prior round's blocking finding is resolved, the change builds, tests are green (41 passed), and i18n is healthy. The remaining items are P2 polish — most notably the duplicate count-line in the no-inner-messages degradation path, which is worth a quick follow-up. Approving.
